apps/passport-server/src/services/generic-issuance/pipelines/PretixPipeline.ts

```diff
@@ -7,37 +52,723 @@ import {
+        secret,
+        checkins
+      } of order.positions) {
+        const product = products.get(item.toString());
+        // The product should always exist, since the validation functions
+        // ensure it. But TypeScript doesn't know that.
+        if (product) {
+          // Try getting email from response to question; otherwise, default to email of purchaser
+          const email = normalizeEmail(attendee_email || order.email);
+
+          // Checkin events can be either "entry" or "exit".
+          // Exits cancel out entries, so we want to find out if the most
+          // recent event was an entry or exit.
+          const checkin = mostRecentCheckinEvent(checkins);
+          // If the most recent event was an entry, the user is checked in
+          const pretix_checkin_timestamp_string =
+            checkin && checkin.type === "entry" ? checkin.datetime : null;
+
+          let pretix_checkin_timestamp: Date | null = null;
+
+          if (pretix_checkin_timestamp_string != null) {
+            try {
+              const parsedDate = Date.parse(
+                pretix_checkin_timestamp_string ?? ""
+              );
+              if (!isNaN(parsedDate)) {
+                pretix_checkin_timestamp = new Date(parsedDate);
+              }
+            } catch (e) {
+              logger(
+                LOG_TAG,
+                "couldn't parse date",
+                pretix_checkin_timestamp_string,
+                e
+              );
+            }
+          }
+
+          tickets.push({
+            email,
+            product,
+            event: eventConfig,
+            full_name: attendee_name || order.name || "", // Fallback since we have a not-null constraint
+            is_consumed: pretix_checkin_timestamp !== null,
+            position_id: id.toString(),
+            secret,
+            pretix_checkin_timestamp
+          });
+        }
+      }
+    }
+    return tickets;
+  }
+
+  private async issuePretixTicketPCDs(
+    req: PollFeedRequest
+  ): Promise<PollFeedResponseValue> {
+    if (!req.pcd) {
+      throw new Error("missing credential pcd");
+    }
+
+    const credential = await SemaphoreSignaturePCDPackage.deserialize(
+      req.pcd.pcd
+    );
+    const feedCredential: FeedCredentialPayload = JSON.parse(
+      credential.claim.signedMessage
+    );
+
+    const serializedEmailPCD = feedCredential.pcd?.pcd;
+    if (!serializedEmailPCD) {
+      throw new Error("missing email pcd");
+    }
+
+    const emailPCD = await EmailPCDPackage.deserialize(serializedEmailPCD);
+    // TODO: verify the email pcd
+    // - that the signature is valid
+    // - that it was signed by the Zupass Server
+
+    const email = emailPCD.claim.emailAddress;
+    const relevantTickets = await this.db.loadByEmail(this.id, email);
+    const ticketDatas = relevantTickets.map((t) =>
+      this.atomToTicketData(t, credential.claim.identityCommitment)
+    );
+
+    // TODO: cache this intelligently
```

Comment:
caching could be important because this server is single-threaded, and signing stuff, especially hundreds of stuff, could have a pretty bad impact on performance